contracts/OpConsensusStateManager.sol

```diff
@@ -0,0 +1,127 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.0;
+
+import './IbcVerifier.sol';
+
+
+// OptimisticConsensusStateManager manages the optimistic consensus
+// states. It keeps track a configurable number of optimistic
+// consensus states. IBC module uses this class to access an
+// optimistic consensus state at a given height to perform membership
+// proof.
+library OptimisticConsensusStateManager {
+    enum OpConsensusStateStatus { NOT_APPLIED, PENDING, TRUSTED }
+
+    struct ManagedOpConsensusState {
+        OptimisticConsensusState state;
+        OpConsensusStateStatus status;
+    }
+
+    struct DataStore {
+        uint32 fraudProofWindowSeconds;
+        uint64 capacity;
+        uint64 size;
+        mapping(uint256 => ManagedOpConsensusState) consensusStates;
+        uint256[] heights;
+    }
+
+    // addOpConsensusState adds an OptimisticConsensusState to
+    // internal store. The newly added op consensus state has a
+    // PENDING status.
+    // 
+    // adding a duplicate OptimisticConsensusState is an idempotent
+    // operation. it doesn't change the fraud proof end time or change
+    // the state of an already trusted OptimisticConsensusState.
+    //
+    // it is an error to update a pending OptimisticConsensusState
+    // with different values. such operation must be done through the
+    // sumitting fraud proof API.
+    //
+    // the status and the fraud proof end time are returned.
+    function addOpConsensusState(DataStore storage self, OptimisticConsensusState calldata state)
```

Comment:
Can you profile the gas costs of these fns?

Would try to simplify logic where we can. On-chain costs can be quite high.